### DISCUSSION TITLE
communicating db problem to the user

### DISCUSSION THREAD

**Discussion by dave42w** [SeaQL/sea-orm#1352] [Q&A]
Hi,

This "works" >>>>>>
pub async fn delete_place(State(state): State<AppState<'_>>, Path(id): Path<String>) -> Redirect {
    let s = state;
    let i = id.parse::<i32>().unwrap();

    let _ = &s.db.transaction::<_, (), DbErr>(|txn| {
        Box::pin(async move {

            let p: Option<place::Model> = Place::find_by_id(i).one(txn).await?;
            if p != None {
                let p: place::ActiveModel = p.unwrap().into();
                p.delete(txn).await?;
            }
            Ok(())
        })
    })
    .await;
    Redirect::to("../../place")
}
<<<<<<

But it doesn't tell the user if the delete fails because someone else deleted the row, other db errors are handled but not reported to the user.

What I would like to do is use HTTP status codes for all the possibilities. I can't quite get my head around this. This is the sort of thing I'm trying:

>>>>
pub async fn delete_place(State(state): State<AppState<'_>>, Path(id): Path<String>) -> Redirect {
    let s = state;
    let i = id.parse::<i32>().unwrap();

    let _ = &s.db.transaction::<_, (), DbErr>(|txn| {
        Box::pin(async move {

            let p: Option<place::Model> = Place::find_by_id(i).one(txn).await?;
            if p != None {
                let p: place::ActiveModel = p.unwrap().into();
                let res = p.delete(txn).await?;
                TODO if rows affected isn't 1 then HTTP status 500
            } else {
              TODO Http status 410 because someone else deleted the row
            }
            Ok(())
        })
    })
    .await;
    TODO only Redirect::to("../../place") if everything went ok,
}
<<<<

Does any of that make sense?

**Comment by billy1624** [comment]
Hey @dave42w, thanks for asking and sorry for the delay!

I think you can make `transaction` closure to return a `HttpErr`.

```rs
let _ = &s.db.transaction::<_, (), HttpErr>(|txn| { // Returns HTTP errors
    Box::pin(async move {

        let p: Option<place::Model> = Place::find_by_id(i).one(txn).await?;
        if p != None {
            let p: place::ActiveModel = p.unwrap().into();
            let res = p.delete(txn).await?.map_err(|e| { // Convert `sea_orm::DbErr` into `HttpErr` });
        } else {
            Err(HttpErr::( ... ))
        }
        Ok(())
    })
})
.await?;
```

**Reply by dave42w** [reply]
Thanks, I'm working my way through this. I haven't done much coding at all for a decade, and am new to rust.
So this integration between SeaOrm and Axum with still unfamiliar Rust syntax is taking me a while.

**Reply by dave42w** [reply]
Hi,
What I'm thinking now is that it would be clearer to move all the db logic into a function that returns a result that is OK or a DbErr.

Then the Axum handler calls that function and decides how to present it to the user. 

That seems simpler than the mixed function I have at the moment that is making multiple db calls and wanting to return to the user from multiple points.
